src/celeritas/UnitTypes.hh

```diff
@@ -0,0 +1,265 @@
+//----------------------------------*-C++-*----------------------------------//
+// Copyright 2023 UT-Battelle, LLC, and other Celeritas developers.
+// See the top-level COPYRIGHT file for details.
+// SPDX-License-Identifier: (Apache-2.0 OR MIT)
+//---------------------------------------------------------------------------//
+//! \file celeritas/UnitTypes.hh
+//! \brief Annotated unit struct definitions for use with Quantity
+//---------------------------------------------------------------------------//
+#pragma once
+
+#include "celeritas_config.h"
+
+#include "Constants.hh"
+#include "Units.hh"
+
+namespace celeritas
+{
+namespace units
+{
+//---------------------------------------------------------------------------//
+//!@{
+//! \name Natural units
+
+struct CLight
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+        return constants::c_light;
+    }
+    static char const* label() { return "c"; }
+};
+
+//! "Natural units" for energy
+struct Mev
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+#if CELERITAS_UNITS == CELERITAS_UNITS_CLHEP
+        return units::megaelectronvolt;
+#else
+        return real_type(1e6) * constants::e_electron * units::volt;
+#endif
+    }
+    static char const* label() { return "MeV"; }
+};
+
+//! "Natural units" for mass
+struct MevPerCsq
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+        return Mev::value() / (constants::c_light * constants::c_light);
+    }
+    static char const* label() { return "MeV/c^2"; }
+};
+
+//! Magnitude of electron charge (positive sign)
+struct EElectron
+{
+    static CELER_CONSTEXPR_FUNCTION real_type value()
+    {
+#if CELERITAS_UNITS == CELERITAS_UNITS_CLHEP
```

Comment:
Why the #if here and not in `Units.hh`? I.e. why not define units::e_electron for the non-CLHEP cases?

Reply:
The comment on the next line was an attempt to explain it but I guess it could use more elaboration. The `units::e_electron` here should be *mathematically* equal to `constants::e_electron`, but might not be *numerically*.  Since we define `units::C` as `1/e` and then define `constants::e` as `e * units::C`, the compiler is doing `(x * (1/x))` which could be one or two ULP off from 1 (i.e. `(x / x)` which is an identity except with nan/zero/inf).

The "constants" should I think for clarity remain constant and not have exceptions based on the unit. The "unit types" here ensures that the unit values for CLHEP are exactly unity.

Reply:
> The "constants" should I think for clarity remain constant

Do you mean that their numerical value are the same independently of the configuration or simply that their are marked `constexpr`?

Reply:
The first :)

Reply:
Then we ought to find a different spelling for the namespace.  For me (and likely others), it says the second :)

Reply:
for example `invariants`

Reply:
in celeritas/Constants.hh in the constant namespace (wikipedia):
> A physical constant, sometimes fundamental physical constant or universal constant, is a [physical quantity](https://en.wikipedia.org/wiki/Physical_quantity) that is generally believed to be both universal in nature and have [constant](https://en.wikipedia.org/wiki/Constant_(mathematics)) value in time. It is distinct from a [mathematical constant](https://en.wikipedia.org/wiki/Mathematical_constant), which has a fixed numerical value, but does not directly involve any physical measurement.

in corecel/Constants.hh in the constant namespace (wikipedia):
>  A mathematical constant is a key [number](https://en.wikipedia.org/wiki/Number) whose value is fixed by an unambiguous definition, often referred to by a special symbol (e.g., an [alphabet letter](https://en.wikipedia.org/wiki/Letter_(alphabet))), or by mathematicians' names to facilitate using it across multiple [mathematical problems](https://en.wikipedia.org/wiki/Mathematical_problem).[[1]](https://en.wikipedia.org/wiki/Mathematical_constant#cite_note-1) Constants arise in many areas of [mathematics](https://en.wikipedia.org/wiki/Mathematics), with constants such as [e](https://en.wikipedia.org/wiki/E_(mathematical_constant)) and [π](https://en.wikipedia.org/wiki/Pi) occurring in such diverse contexts as [geometry](https://en.wikipedia.org/wiki/Geometry), [number theory](https://en.wikipedia.org/wiki/Number_theory), [statistics](https://en.wikipedia.org/wiki/Statistics), and [calculus](https://en.wikipedia.org/wiki/Calculus).

Am I missing something???

Reply:
And in `celeritas/Units.hh` in the `units` namespace:
> A system of units of measurement, also known as a system of units or system of measurement, is a collection of [units of measurement](https://en.wikipedia.org/wiki/Units_of_measurement) and rules relating them to each other. Systems of measurement have historically been important, regulated and defined for the purposes of [science](https://en.wikipedia.org/wiki/Science) and [commerce](https://en.wikipedia.org/wiki/Commerce). Instances in use include the [International System of Units](https://en.wikipedia.org/wiki/International_System_of_Units) or SI (the modern form of the [metric system](https://en.wikipedia.org/wiki/Metric_system)), the [British imperial system](https://en.wikipedia.org/wiki/British_imperial_system), and the [United States customary system](https://en.wikipedia.org/wiki/United_States_customary_system).

Reply:
It is unqualified so maybe it [refers to a collection of:](https://en.cppreference.com/w/cpp/language/cv)
```
Appear in any type specifier, including decl-specifier-seq of [declaration grammar](https://en.cppreference.com/w/cpp/language/declarations), to specify constness or volatility of the object being declared or of the type being named.

    const - defines that the type is constant.
```

Taking from the quotes above, maybe `physical_constant` (or `math_constant`) :)

Reply:
More importantly it does not indicates (for better or worse) where the `constant` numerical value are configuration dependent (here depending on the units selected) or not.  Or maybe it does not matter.

Reply:
They're always constant and in the native unit system.

Reply:
Given this:
(a)
```
celeritas_setup_option(CELERITAS_UNITS CGS)
celeritas_setup_option(CELERITAS_UNITS SI)
celeritas_setup_option(CELERITAS_UNITS CLHEP)
celeritas_define_options(CELERITAS_UNITS
  "Native unit system for Celeritas")
 ```

(b)
 > They're always constant and in the native unit system.
 
(c)
> Do you mean that their numerical value are the same independently of the configuration or ...
> The first :)

Either the meaning of 'native' changes between (a) and (b) or (c) is incorrect (if the constant are expressed in native unit system and the native unit system change with the configuration ( (a) says so) then the numerical value changes with the configuration.

Reply:
I see what you mean; the numerical value of `c_light` changes depending on the compile option for choosing the native unit system. They are always consistent with the "native" unit system, and they are fundamental physical constants.

Reply:
Alright. So `celeritas::constants` currently contains universal constant both mathematical (and unit-less) (eg. `pi`) and physical (and with units) (eg. `c_light`).   A bit awkward with fine.  Another awkward-ness is that `orange/Constants.hh` declares things (somewhat unexpectedly) in the namespace `celeritas::constants`.

Reply:
Yeah the ORANGE constants have a "todo: move to corecel" there. I think it's OK for both the math and physical constants to coexist. In SI or (a given derivative unit system) the main physical constants have an exactly specified numerical value, and the remaining (measured) physical constants have some fuzziness on them.